sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -679,6 +679,8 @@ object RemoveNoopUnion extends Rule[LogicalPlan] {
       d.withNewChildren(Seq(simplifyUnion(u)))
     case d @ Deduplicate(_, u: Union) =>
       d.withNewChildren(Seq(simplifyUnion(u)))
+    case d @ DeduplicateWithinWatermark(_, u: Union) =>
```

Comment:
Rather than making this a separate logical node, we can we make the new behavior an option in Deduplicate node? That way we don't need to distinguish them in the implementation except in couple of places.

Reply:
dropDuplicates() produces different output between batch and streaming too (because batch does not drop late records). 
As much as possible it is better to keep the API same between batch and streaming.

Reply:
No what I meant is, if we assume the perfect watermark on streaming side, streaming query is designed to produce the same output with batch one. dropDuplicates() is no exception.

Arguably, dropDuplicatesWithinWatermark is an exception because they are not only required to reason about "lateness", but also required to reason about max time duration on duplicated events. Would it be safe to just change read to readStream / write to writeStream and vice versa? Mostly yes for existing API, but maybe not for this API.

Reply:
There is no perfect watermark, and any guarantees that apply only with "perfect watermark" are no guarantees at all. 
I don't think semantic deferences require disabling batch.

Reply:
> I don't think semantic deferences require disabling batch.

+1. I don't see why this should be different than the existing `dropDuplicates`.

Reply:
I'm OK with that. Let's say batch query has infinite delay threshold of watermark. That will explain the behavior of batch query for this operator.

Reply:
@zsxwing one more thing to consider: Shall we extend existing 'DropDuplicate' Node to support this use case rather than introducing another logical node 'DropDuplicatesWithinWatermark'. It does not change the API, just the implementation. I am suggesting adding this option to exiting node.

Reply:
> Shall we extend existing 'DropDuplicate' Node to support this use case rather than introducing another logical node 'DropDuplicatesWithinWatermark'.

I don't have a strong opinion on this. Will let you two to decide 😄

Reply:
Yeah, went with the two node option. @HeartSaVioR reused a lot of the code.